Add Fiber::ExecutionContext::Parallel#resize

ysbaddaden opened the pull request:
Allows to dynamically resize a Parallel execution context to a new maximum parallelism.

**It can grow**: merely adds more schedulers, which may eventually start more threads.

**It can shrink**: immediately removes any overflow schedulers and tell them to shutdown. The actual shutdown is cooperative, a scheduler won't notice the shutdown until it's current fiber is rescheduled. Parallelism will eventually be reduced.

Follow up to #15936 and #15946.

ysbaddaden commented on src/fiber/execution_context.cr:
```diff
@@ -8,25 +8,25 @@ require "./execution_context/*"
 # Applications can create any number of execution contexts in parallel. These
-# contexts are isolated but they can communicate with the usual thread-safe
-# synchronization primitives (e.g. `Channel`, `Mutex`).
+# contexts are isolated but they can communicate with the usual synchronization
```
```suggestion
# contexts are isolated but they can communicate using the usual synchronization
```

ysbaddaden commented on src/fiber/execution_context/parallel/scheduler.cr:
```diff
@@ -29,13 +29,23 @@ module Fiber::ExecutionContext
       end
 
+      # :nodoc:
+      enum Shutdown
+        NO  = 0
+        NOW
+      end
```
`Shutdown::NO` is weird. If someone has a better suggestion...

straight-shoota replied:
Do we even need an enum type for such an internal detail? Why not make it a boolean instead?

ysbaddaden replied:
We're planning to tell an execution context to shutdown and the behavior will be different to shutdown _now_ (drain your local queue and stop immediately) and will instead be _when nothing left to do_ (stop instead of parking) for example.

straight-shoota commented on src/fiber/execution_context/runnables.cr:
```diff
@@ -75,25 +75,54 @@ module Fiber::ExecutionContext
-      # append fiber to the batch
+      # append fiber to the batch and push to global queue
       batch.to_unsafe[n] = fiber
+      push_to_global_queue(batch.to_unsafe, n + 1)
```
**note:** The previous implementation was using the wrapping `&+` operator. Is this change intended?

ysbaddaden commented:
I dropped the `Shutdown::NO` value and replaced with an invalid value until it's set to `Shutdown::NOW`. I kept it since `@shutdown.now?` reads well, but maybe we could drop it, `@shutdown_now` would read just as well. That depends on the cooperative shutdown.

I also fixed an issue (must reset `@parked`) and added an early abort when spinning (no need to spin). I kept the non-blocking evloop run, which doesn't hurt... it might be preferable to enqueue to the global queue directly, though :thinking:

ysbaddaden commented on spec/std/fiber/execution_context/parallel_spec.cr:
```diff
@@ -41,4 +42,43 @@ describe Fiber::ExecutionContext::Parallel do
+
+      ctx.spawn do
+        while running.get(:relaxed)
+          sleep(100.microseconds)
```
Let's go for 10us as it would mean a quicker resume time and more opportunities for messing up with the resize. It doesn't affect the spec run time, though.

```suggestion
          sleep(10.microseconds)
```

ysbaddaden replied:
Explainer: The point of the spec is to have fibers running & sleeping in the schdeulers, so we can test that resize/drain is working correctly (no segfault, no lost fiber).

I tried with sleeps from 0 to 100us and the spec runtime didn't change. I guess even waiting for 100us gets hidden by the dozen calls to `#resize` that must start/stop schedulers/thread and steal/drain/run the evloop.
